system/lib/pthread/pthread_create.c

```diff
@@ -118,44 +145,50 @@ int __pthread_create(pthread_t* restrict res,
 
   new->locale = &libc.global_locale;
 
-  // Allocate memory for thread-local storage and initialize it to zero.
+  // tsd slots are at the very end of the block
+  unsigned char* tsd = block + size - __pthread_tsd_size;
   if (__pthread_tsd_size) {
-    new->tsd = malloc(__pthread_tsd_size);
-    memset(new->tsd, 0, __pthread_tsd_size);
+    new->tsd = (void*)tsd;
   }
 
-  new->detach_state = DT_JOINABLE;
-
-  if (attrp && attrp != __ATTRP_C11_THREAD) {
-    if (attrp->_a_detach) {
-      new->detach_state = DT_DETACHED;
-    }
-    new->stack_size = attrp->_a_stacksize;
-    new->stack = (void*)attrp->_a_stackaddr;
+  if (attr._a_detach) {
+    new->detach_state = DT_DETACHED;
+  } else {
+    new->detach_state = DT_JOINABLE;
+  }
+  new->stack_size = attr._a_stacksize;
+  // musl stores top of the stack in pthread_t->stack (i.e. the high
+  // end from which it grows down).
+  if (attr._a_stackaddr) {
+    new->stack = (void*)attr._a_stackaddr;
   } else {
-    new->stack_size = _emscripten_default_pthread_stack_size();
+    new->stack = (void*)((intptr_t)tsd & ~(STACK_ALIGN-1));
   }
 
-  if (!new->stack) {
-    char* stackBase = memalign(STACK_ALIGN, new->stack_size);
-    // musl stores top of the stack in pthread_t->stack (i.e. the high
-    // end from which it grows down).
-    new->stack = stackBase + new->stack_size;
-    new->stack_owned = 1;
+  if (__builtin_wasm_tls_size()) {
+    struct pthread * struct_end = new + 1;
+    new->tls_base = (void*)ROUND_UP((intptr_t)struct_end, __builtin_wasm_tls_align());
```

Comment:
so, `tls_base` now points past the bounds of `new`? But that's ok because we allocated more memory than just the size of a `struct pthread`?

Reply:
Indeed.. we allocated enough for struct + tls + stack + tsd (see the command above for the layout of this allocation).